packages/rpc/index.js

```diff
@@ -80,16 +100,75 @@ class SlashtagsRPC extends EventEmitter {
     return rpc
   }
 
+  /**
+   * @param {Uint8Array | string} key
+   */
+  async _rpcFromSwarm (key) {
+    /** @type {Uint8Array} */
+    const publicKey =
+      typeof key === 'string'
+        ? key.startsWith('slash')
+          ? SlashURL.parse(key).key
+          : SlashURL.decode(key)
+        : key
+
+    // duplicate logic from Hyperswarm's internal connections management, but swarm._allConnections is not a public API, so we can't rely on it.
+    const existingConnection = this._allConnections.get(publicKey)
+    if (existingConnection) {
+      // this.setup() is idempotent
+      return this.setup(existingConnection)
+    }
+
+    this._swarm.joinPeer(publicKey)
+
+    // Await for connecting to every seeder for unit testing convenience
+    return new Promise((resolve) => {
+      /** @param {import('protomux-rpc')} rpc */
+      const cleanupAndResolve = (rpc) => {
+        this._swarm.removeListener('connection', onConnection)
+        resolve(rpc)
+      }
+
+      /** @param {SecretStream} connection */
+      const onConnection = (connection) => {
+        if (b4a.equals(connection.remotePublicKey, publicKey)) {
+          this._allConnections.set(publicKey, connection)
+          connection.once('close', () => {
+            this._allConnections.delete(publicKey)
+          })
+
+          const rpc = this.setup(connection)
+          cleanupAndResolve(rpc)
+        }
+      }
+
+      // Timeout to avoid blocking this.ready()
+      setTimeout(cleanupAndResolve, CONNECT_TO_SEEDER_TIMEOUT)
```

Comment:
`CONNECT_TO_SEEDER_TIMEOUT` - is more like `CONNECT_TO_SEEDER__DELAY`  rather than timeout, no?

Reply:
No this is supposed to put a timeout on the `_rpcFromSwarm` promise, because `hyperswarm.joinPeer()` doesn't really return a promise or timeouts, so we have to wait for the connection, but it is also important for all functions to fail in a maximum timeout, otherwise it would hang forever if the target isn't dialable.

It should be renamed to `CONNECT_TIMEOUT` though, I just copied it from `core-data` without thinking about the name.